Enhance setup.py cmake errors reporting (#111)

caliuf opened the pull request:
Handle possible cmake errors by displaying them to the user.

Referenced in #111

caliuf committed:
Enhance setup.py cmake errors reporting

McSinyx commented:
Thank you for submitting this patch!  Since I'm a bit lazy, could you please show us the output of `pip install .` when an error occurs.  A screenshot might be preferable if it fits in one screen, mainly for me to see the colors.

I'll review this somewhere within tomorrow.  Until then, please add your name to the copyright header of the file you're modifying and fix the linting errors raised by flake8.

caliuf commented:
Here's how it looks:
![setup_error](https://user-images.githubusercontent.com/10155161/88815141-4716fd80-d1bb-11ea-80d2-04a2a56f8397.png)

I'll print on the stderr instead of stdout by the way (but this doesn't change output colors in my system)

caliuf committed:
Update

caliuf committed:
Update

caliuf commented:
I updated the patch. With the most recent pip version the output is the same in the screenshot, aside that it's all red.

Check if I changed copyright notice correctly.

Huy-Ngo commented on setup.py:
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/env python3
 # setup script
-# Copyright (C) 2019, 2020  Nguyễn Gia Phong
+# Copyright (C) 2019, 2020  Nguyễn Gia Phong, Francesco Caliumi
```
We have copyright notice for each author on a separate line. Yours would look like this:

```python
# Copyright (C) 2020  Francesco Caliumi
```

Huy-Ngo commented on setup.py:
```diff
@@ -57,8 +58,19 @@ def finalize_options(self) -> None:
+            cmake = run(['cmake', '.'],
+                        check=True, stdout=DEVNULL, stderr=PIPE,
+                        cwd=self.build_temp, universal_newlines=True)
+        except CalledProcessError as e:
+            print(f"\n\ncmake: Error occurred!\n\n"
+                  f"(Standard output)\n"
+                  f"{e.stdout}\n\n"
+                  f"(Standard error)\n"
+                  f"{e.stderr}\n",
+                  file=sys.stderr)
```
In this project, we prefer to use single quotation mark `'` to double quotation mark `"`

Huy-Ngo reviewed (changes requested):
Thank you for submitting this patch. I just have a few comments on the copyright notice and code style.

McSinyx commented on setup.py:
```diff
@@ -57,8 +58,19 @@ def finalize_options(self) -> None:
+                        check=True, stdout=DEVNULL, stderr=PIPE,
+                        cwd=self.build_temp, universal_newlines=True)
+        except CalledProcessError as e:
+            print(f"\n\ncmake: Error occurred!\n\n"
+                  f"(Standard output)\n"
+                  f"{e.stdout}\n\n"
+                  f"(Standard error)\n"
+                  f"{e.stderr}\n",
+                  file=sys.stderr)
```
The following seems to be sufficient:
```suggestion
            print(e.stderr, file=sys.stderr)
```

caliuf replied:
@McSinyx I'm unfamiliar with cmake and if stdout could help to clarify what's going on if an error arise, so it's ok for me. 

Giving no clues about what the cmake error output is could create confusion in a distracted user, let me now if you are sure about this

McSinyx replied:
As shown in your screenshot, CMake actually prints `CMake Error at ...` when there's a problem and such output will always be delivered to stderr.  In addition the script to be executed here also prints to stderr, so there's never any output to stdout.

McSinyx commented on setup.py:
```diff
@@ -57,8 +58,19 @@ def finalize_options(self) -> None:
+                  f"(Standard error)\n"
+                  f"{e.stderr}\n",
+                  file=sys.stderr)
+            sys.exit(1)
+
```
I'd prefer a re-raise here, since it'd contain the error message which has the exit code of CMake.  
```suggestion
            raise
```

caliuf replied:
I did this because the stacktrace is very misleading on what the error really is, but I see your point I'll change that

McSinyx replied:
You got a point about the traceback, so how about `raise e from None`?

caliuf committed:
Update

caliuf committed:
Update

caliuf commented:
I updated the PR with your suggestions. This is what it looks like now (pip 20.2, python 3.7):

![setup-error](https://user-images.githubusercontent.com/10155161/88909698-be9c6980-d25b-11ea-8417-b413c4dcf623.png)

McSinyx commented on setup.py:
```diff
@@ -17,6 +18,7 @@
 # You should have received a copy of the GNU Lesser General Public License
 # along with palace.  If not, see <https://www.gnu.org/licenses/>.
 
+import sys
```
flake8 is complaining about this

caliuf commented:
One of the appveyor.com jobs died for a timeout, I think it's mergeable

Huy-Ngo reviewed (approved):
Looks good to me

McSinyx commented:
@caliuf, did you see my comment at https://github.com/McSinyx/palace/pull/112#discussion_r462889554?

caliuf commented:
@McSinyx I didn't know the `raise X from Y` syntax, and I can hardly find any mention of it on the web, what does it do exactly? 
At first glance it seems that Y should be an (exception) object, a form of "X error happened while processing Y", but I can't see how it can help in this case.

I've tried it and the output seems exactly the same

caliuf committed:
Update

McSinyx reviewed (approved):
You're right, [raise x from y](https://docs.python.org/3/library/exceptions.html) doesn't help the traceback if there's only one exception raised through the handling process.  I'll go ahead merging this now and iterate on it later.  Thanks for your patience, @caliuf!
